### PR TITLE
Ensure that expired/removed downtimes/comments are correctly updated in DB IDO

### PR DESCRIPTION
**Pull request by Crunsher**
This adds an additional index to the downtimehistory and updates the
dbevents in a way that it uses that index.

Refs #5458

### CODE REVIEW

**Comment by dnsmichi** [comment]
The upgrade SQL files are 2.6.1, not 2.7.1

**Comment by dnsmichi** [comment]
Please also raise `IDO_CURRENT_SCHEMA_VERSION` to `1.14.3`, but leave `IDO_COMPAT_SCHEMA_VERSION` as is in https://github.com/Icinga/icinga2/blob/master/lib/db_ido/dbconnection.hpp#L32

**Comment by Crunsher** [comment]
This fix also seems not to be working as expected

```
MariaDB [icinga]> select count(name) from icinga_downtimehistory where is_in_effect = 1;
+-------------+
| count(name) |
+-------------+
|          11 |
+-------------+
```

The index is used
```
MariaDB [icinga]> explain select * from icinga_downtimehistory WHERE object_id = 0 AND entry_time = "" AND instance_id = 0 and scheduled_start_time = 0 and scheduled_end_time = 0 AND name = "";+------+-------------+------------------------+------+-------------------------------------------------------+-------------+---------+-------------------+------+-------------+
| id   | select_type | table                  | type | possible_keys                                         | key         | key_len | ref               | rows | Extra       |
+------+-------------+------------------------+------+-------------------------------------------------------+-------------+---------+-------------------+------+-------------+
|    1 | SIMPLE      | icinga_downtimehistory | ref  | instance_id,sla_idx_dohist,idx_downtimehistory_delete | instance_id | 23      | const,const,const |    1 | Using where |
+------+-------------+------------------------+------+-------------------------------------------------------+-------------+---------+-------------------+------+-------------+
1 row in set, 1 warning (0.00 sec)
```

But somehow there are still mismatches

**Comment by Crunsher** [comment]
I updated the wrong query ><

Now scheduleddowntimes also has an index on the same columns and uses them for deletion.
We should also investigate if this problem also concerns comments

**Comment by Crunsher** [comment]
Added @dnsmichi as reviewer. Wouldn't want to fuck up schema changes ^_^

File: lib/db_ido/dbevents.cpp [hunk 1]
```diff
@@ -407,10 +407,10 @@ void DbEvents::RemoveCommentInternal(std::vector<DbQuery>& queries, const Commen
 	query1.Type = DbQueryDelete;
 	query1.Category = DbCatComment;
 	query1.WhereCriteria = new Dictionary();
-	query1.WhereCriteria->Set("internal_comment_id", comment->GetLegacyId());
 	query1.WhereCriteria->Set("object_id", checkable);
-	query1.WhereCriteria->Set("comment_time", DbValue::FromTimestamp(entry_time));
+	query1.WhereCriteria->Set("entry_time", DbValue::FromTimestamp(entry_time));
 	query1.WhereCriteria->Set("instance_id", 0); /* DbConnection class fills in real ID */
+	query1.WhereCriteria->Set("name", comment->GetName());
```
Comment (dnsmichi) [inline lib/db_ido/dbevents.cpp]: We agreed on not using `name` for 2.7.1, but only for 2.8.0 then. Please update your patches.

File: lib/db_ido/dbevents.cpp [hunk 2]
```diff
@@ -430,8 +430,9 @@ void DbEvents::RemoveCommentInternal(std::vector<DbQuery>& queries, const Commen
 	query2.WhereCriteria = new Dictionary();
 	query2.WhereCriteria->Set("internal_comment_id", comment->GetLegacyId());
 	query2.WhereCriteria->Set("object_id", checkable);
-	query2.WhereCriteria->Set("comment_time", DbValue::FromTimestamp(entry_time));
+	query1.WhereCriteria->Set("entry_time", DbValue::FromTimestamp(entry_time));
 	query2.WhereCriteria->Set("instance_id", 0); /* DbConnection class fills in real ID */
+	query1.WhereCriteria->Set("name", comment->GetName());
```
Comment (dnsmichi) [inline lib/db_ido/dbevents.cpp]: Same, name in the where condition.

File: lib/db_ido/dbevents.cpp [hunk 3]
```diff
@@ -584,10 +585,13 @@ void DbEvents::RemoveDowntimeInternal(std::vector<DbQuery>& queries, const Downt
...
 	query1.Category = DbCatDowntime;
 	query1.WhereCriteria = new Dictionary();
+
 	query1.WhereCriteria->Set("object_id", checkable);
-	query1.WhereCriteria->Set("internal_downtime_id", downtime->GetLegacyId());
 	query1.WhereCriteria->Set("entry_time", DbValue::FromTimestamp(downtime->GetEntryTime()));
 	query1.WhereCriteria->Set("instance_id", 0); /* DbConnection class fills in real ID */
+	query1.WhereCriteria->Set("scheduled_start_time", DbValue::FromTimestamp(downtime->GetStartTime()));
+	query1.WhereCriteria->Set("scheduled_end_time", DbValue::FromTimestamp(downtime->GetEndTime()));
+	query1.WhereCriteria->Set("name", downtime->GetName());
```
Comment (dnsmichi) [inline lib/db_ido/dbevents.cpp]: name again.

File: lib/db_ido/dbevents.cpp [hunk 4]
```diff
@@ -612,9 +616,11 @@ void DbEvents::RemoveDowntimeInternal(std::vector<DbQuery>& queries, const Downt
 
 	query3.WhereCriteria = new Dictionary();
 	query3.WhereCriteria->Set("object_id", checkable);
-	query3.WhereCriteria->Set("internal_downtime_id", downtime->GetLegacyId());
 	query3.WhereCriteria->Set("entry_time", DbValue::FromTimestamp(downtime->GetEntryTime()));
 	query3.WhereCriteria->Set("instance_id", 0); /* DbConnection class fills in real ID */
+	query3.WhereCriteria->Set("scheduled_start_time", DbValue::FromTimestamp(downtime->GetStartTime()));
+	query3.WhereCriteria->Set("scheduled_end_time", DbValue::FromTimestamp(downtime->GetEndTime()));
+	query3.WhereCriteria->Set("name", downtime->GetName());
```
Comment (dnsmichi) [inline lib/db_ido/dbevents.cpp]: name.

**Comment by Crunsher** [comment]
I'll leave it be until we have clarification on how to proceed with this

**Comment by dnsmichi** [comment]
We're going to use the partial index here for 2.7.1.

**Comment by dnsmichi** [comment]
## Index Tests

```
MariaDB [icinga]> show create table icinga_downtimehistory;
+------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table                  | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
+------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| icinga_downtimehistory | CREATE TABLE `icinga_downtimehistory` (
  `downtimehistory_id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `instance_id` bigint(20) unsigned DEFAULT '0',
  `downtime_type` smallint(6) DEFAULT '0',
  `object_id` bigint(20) unsigned DEFAULT '0',
  `entry_time` timestamp NULL DEFAULT NULL,
  `author_name` varchar(64) CHARACTER SET latin1 DEFAULT '',
  `comment_data` text CHARACTER SET latin1,
  `internal_downtime_id` bigint(20) unsigned DEFAULT '0',
  `triggered_by_id` bigint(20) unsigned DEFAULT '0',
  `is_fixed` smallint(6) DEFAULT '0',
  `duration` bigint(20) DEFAULT '0',
  `scheduled_start_time` timestamp NULL DEFAULT NULL,
  `scheduled_end_time` timestamp NULL DEFAULT NULL,
  `was_started` smallint(6) DEFAULT '0',
  `actual_start_time` timestamp NULL DEFAULT NULL,
  `actual_start_time_usec` int(11) DEFAULT '0',
  `actual_end_time` timestamp NULL DEFAULT NULL,
  `actual_end_time_usec` int(11) DEFAULT '0',
  `was_cancelled` smallint(6) DEFAULT '0',
  `is_in_effect` smallint(6) DEFAULT '0',
  `trigger_time` timestamp NULL DEFAULT NULL,
  `name` text CHARACTER SET latin1,
  `endpoint_object_id` bigint(20) DEFAULT NULL,
  PRIMARY KEY (`downtimehistory_id`),
  UNIQUE KEY `instance_id` (`instance_id`,`object_id`,`entry_time`,`internal_downtime_id`),
  KEY `sla_idx_dohist` (`object_id`,`actual_start_time`,`actual_end_time`)
) ENGINE=InnoDB AUTO_INCREMENT=16157 DEFAULT CHARSET=utf8 COMMENT='Historical scheduled host and service downtime' |
+------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

MariaDB [icinga]> explain update icinga_downtimehistory set was_cancelled=1, actual_end_time='2017-09-19 11:59:39', actual_end_time_usec=185877, is_in_effect=0 where object_id=5144 and entry_time='2017-09-19 11:59:33' and instance_id=1 and internal_downtime_id=14042;
+------+-------------+------------------------+-------+----------------------------+-------------+---------+------+------+-------------+
| id   | select_type | table                  | type  | possible_keys              | key         | key_len | ref  | rows | Extra       |
+------+-------------+------------------------+-------+----------------------------+-------------+---------+------+------+-------------+
|    1 | SIMPLE      | icinga_downtimehistory | range | instance_id,sla_idx_dohist | instance_id | 32      | NULL |    1 | Using where |
+------+-------------+------------------------+-------+----------------------------+-------------+---------+------+------+-------------+
1 row in set (0.00 sec)

MariaDB [icinga]> explain update icinga_downtimehistory set was_cancelled=1, actual_end_time='2017-09-19 11:59:39', actual_end_time_usec=185877, is_in_effect=0 where object_id=5144 and entry_time='2017-09-19 11:59:33' and instance_id=1 and name='many-test-23!icinga!mbmif.int.netways.de-1505815173-14040';
+------+-------------+------------------------+-------+----------------------------+-------------+---------+------+------+-------------+
| id   | select_type | table                  | type  | possible_keys              | key         | key_len | ref  | rows | Extra       |
+------+-------------+------------------------+-------+----------------------------+-------------+---------+------+------+-------------+
|    1 | SIMPLE      | icinga_downtimehistory | range | instance_id,sla_idx_dohist | instance_id | 23      | NULL |    1 | Using where |
+------+-------------+------------------------+-------+----------------------------+-------------+---------+------+------+-------------+
1 row in set (0.01 sec)
```

Conclusion: The code fixes for 2.7.1 should suffice here using the partial index.

**Comment by dnsmichi** [comment]
This should also include changes for the status tables and unique constraints (to be evaluated) where the unique constraints uses `name` as varchar instead of the downtime id.

**Comment by Crunsher** [comment]
Fixed a little copy-paste error and rebased against the master

**Comment by lippserd** [comment]
Hi,

I moved the 2.7.1.sql migration scripts to 2.8.0.sql and removed the 4 instance_id indices.

Cheers,
Eric

**Comment by lippserd** [comment]
Hi Jean,

There are several places in the code where we use `instance_id` as extra where criteria for the downtime and comment queries (trigger downtime, removal queries). This is not necessary as we already have `object_id` and ìnstance_id` is not covered by the new index. Please find and fix all places to match the new index. Further, `internal_comment_id` and `internal_downtime_id` are still in use as where criteria for some queries. Do we have to fix that as well?

Cheers,
Eric

**Comment by Crunsher** [comment]
You are right, using internal_comment/downtime_id could cause the same issue when adding downtimes. And instance_id is irrelevant thanks to object_id

**Comment by dnsmichi** [comment]
Note: The idea of changing name into a varchar and adding it to the index is gone. Instead we rely on the partial index. The condition that you won't match the index is near to zero, when you have created multiple downtimes/comments in the same second, and they are to be deleted/updated.

One thing I'll add - compat schema must be the current one, since we want to avoid possible workarounds and missing indexes here. Users must carefully run the upgrade.

**Comment by dnsmichi** [comment]
Tested it both with MySQL and PostgreSQL again. Since this patch only solves the regression, the real problem was already fixed in v2.7.1 with downtimehistory updates.

**Comment by dnsmichi** [comment]
Thanks. Good job everyone on a difficult topic 💪  👍

**Pull request merged**